src/tsf/Implementation.cpp

```diff
@@ -500,7 +499,7 @@ void Implementation::_doCompositionUpdate(TfEditCookie ec)
                 ULONG len = bufCap;
                 THROW_IF_FAILED(range->GetText(ec, TF_TF_MOVESTART, buf, len, &len));
 
-                if (!composing && firstRange)
+                if (!composing)
                 {
                     finalizedString.append(buf, len);

```

Comment:
And this doesn't cause us to re-commit repeated text for *other* languages?

Reply:
Hmm I can make it so that any series of non-composing runs at the start all gets finalized. That's probably the best trade-off.

Reply:
Did I actually find an issue, or is this defensive coding?

What i'm really asking is: why did this boolean exist? Are we reintroducing a different bug?

Reply:
Only defensive coding. I added it as a gut feeling that removing a finalized part between two active/ongoing compositions would be bad. I'll keep the defense but make it so that it can remove >1 composition run from the start of the whole thing.